### PR TITLE
Use records for cross arguments.

### LINKED ISSUE #1585: Use records for cross' transition function
We should do the same as in #1580 for `cross`. Currently, the type is
```
(?id : string, ?active : bool, ?conservative : bool,
 ?duration : {float}, ?minimum : float,
 ?override_duration : string, ?width : float,
 ((float, float, [string * string], [string * string],
   source(audio=pcm('a), video='b, midi='c),
   source(audio=pcm('a), video='b, midi='c)) ->
  source(audio=pcm('a), video='b, midi='c)),
 source(audio=pcm('a), video='b, midi='c)) ->
source(audio=pcm('a), video='b, midi='c)
```
which is quite complicated: we want the crossing function to take two sources as argument (not 6 arguments), each one being decorated by fields indicating properties such as metadata, power, etc.

### PR DESCRIPTION
**Pull request by smimram**
Fixes #1585.

### CODE REVIEW

File: src/operators/cross.ml [hunk 1]
```diff
@@ -374,12 +374,18 @@ class cross ~kind (s : source) ~cross_length ~override_duration ~rms_width
...
-                  ("", Lang.float db_before);
-                  ("", Lang.float db_after);
-                  ("", Lang.metadata before_metadata);
-                  ("", Lang.metadata after_metadata);
-                  ("", Lang.source a);
-                  ("", Lang.source b);
+                  ( "",
+                    Lang.meth (Lang.source a)
+                      [
+                        ("power", Lang.float db_before);
```
Comment (toots) [inline src/operators/cross.ml:380]: I think `power` is a bit vague. `dbLevel` would be better.

**Comment by toots** [comment]
> We have a typing error there for good reasons in `fades.liq`. Basically, the transition function does not consider power which thus does not show up in the type.
> 
> We could work around by adding
> 
> ```
> ignore(a.power)
> ignore(b.power)
> ```
> 
> in the transition function, but the standard users will not find this by themselves. The problem here is that our typing system does not allow to say that we take as input a type of the form
> 
> ```
> source.{metadata : metadata, ...}
> ```
> 
> (with the `...` indicating that we might have more...)

If I remember correctly, adding this would be pretty complex right?

**Comment by smimram** [comment]
> If I remember correctly, adding this would be pretty complex right?

This should be doable, but definitely post-2.0. I have some ideas that I want to try, but this would mean heavily re-thinking the type system once again... :)

**Comment by toots** [comment]
> > If I remember correctly, adding this would be pretty complex right?
> 
> This should be doable, but definitely post-2.0. I have some ideas that I want to try, but this would mean heavily re-thinking the type system once again... :)

I hear you. Should we revert https://github.com/savonet/liquidsoap/issues/1580?

**Comment by smimram** [comment]
> > > If I remember correctly, adding this would be pretty complex right?
> > 
> > 
> > This should be doable, but definitely post-2.0. I have some ideas that I want to try, but this would mean heavily re-thinking the type system once again... :)
> 
> I hear you. Should we revert #1580?

Nope, I had a thought about that: we can represent `{x:int,...}` because it is a `'a` decorated with a field `x:int` (i.e. `'a.{x:int}`). However we cannot represent a type such as `source.{x:int,...}`. So for pure records it's fine, but not for decorated values. For cross we could take 4 four arguments: a source + a record of parameter (for each of the two sources).

**Comment by toots** [comment]
> > > > If I remember correctly, adding this would be pretty complex right?
> > > 
> > > 
> > > This should be doable, but definitely post-2.0. I have some ideas that I want to try, but this would mean heavily re-thinking the type system once again... :)
> > 
> > 
> > I hear you. Should we revert #1580?
> 
> Nope, I had a thought about that: we can represent `{x:int,...}` because it is a `'a` decorated with a field `x:int` (i.e. `'a.{x:int}`). However we cannot represent a type such as `source.{x:int,...}`. So for pure records it's fine, but not for decorated values. For cross we could take 4 four arguments: a source + a record of parameter (for each of the two sources).

Neat!

**Pull request merged**